packages/fiber/tests/utils.test.ts

```diff
@@ -97,3 +113,305 @@ describe('is', () => {
+    expect(child.previousAttach).toBe(undefined)
+  })
+
+  it('should create array when using array-index syntax', () => {
+    const parent = prepare({ prop: null }, null!, '', {})
+    const child = prepare({}, null!, '', { attach: 'prop-0' })
+
+    attach(parent, child)
+    expect(parent.object.prop).toStrictEqual([child.object])
+    expect(child.previousAttach).toBe(undefined)
+
+    detach(parent, child)
+    expect((parent.object.prop as unknown as Array<never>).length).toBe(1)
+    expect((parent.object.prop as unknown as Array<never>)[0]).toBe(undefined)
+    expect(child.previousAttach).toBe(undefined)
+  })
+})
+
+describe('diffProps', () => {
+  it('should filter changed props', () => {
+    const oldProps = { foo: true }
+    const newProps = { foo: true, bar: false }
+
+    const filtered = diffProps(oldProps, newProps)
+    expect(filtered).toStrictEqual({ bar: false })
+  })
+
+  it('should pick removed props for HMR', () => {
+    const oldProps = { foo: true, bar: false }
+    const newProps = { foo: true }
+
+    const filtered = diffProps(oldProps, newProps, true)
+    expect(filtered).toStrictEqual({ bar: DEFAULT + 'remove' })
+  })
+
+  it('should filter reserved props without accessing them', () => {
+    const get = jest.fn()
+    const set = jest.fn()
+
+    const props = { foo: true }
+    const filtered = diffProps(
+      {},
+      RESERVED_PROPS.reduce((acc, key) => ({ ...acc, [key]: { get, set } }), props),
+    )
+
+    expect(filtered).toStrictEqual(props)
+    expect(get).not.toBeCalled()
+    expect(set).not.toBeCalled()
+  })
+})
+
+describe('applyProps', () => {
+  it('should apply props to foreign objects', () => {
+    const target = new THREE.Object3D()
+    expect(() => applyProps(target, {})).not.toThrow()
+  })
+
+  it('should filter reserved props without accessing them', () => {
+    const get = jest.fn()
+    const set = jest.fn()
+
+    const props = { foo: true }
+    const target = {}
+    applyProps(
+      target,
+      RESERVED_PROPS.reduce((acc, key) => ({ ...acc, [key]: { get, set } }), props),
+    )
+
+    expect(target).toStrictEqual(props)
+    expect(get).not.toBeCalled()
+    expect(set).not.toBeCalled()
+  })
+
+  it('should reset removed props for HMR', () => {
+    const target = new THREE.Object3D()
+    const instance = prepare(target, { getState: () => null! } as any, '', {})
+    prepare(target.scale, null!, '', { args: [5, 5, 5] })
+    target.position.setScalar(10)
+
+    // Recreate from args
+    applyProps(target, { scale: DEFAULT + 'remove' })
+    expect(target.scale.toArray()).toStrictEqual([5, 5, 5])
+
+    // Recreate from scratch
+    applyProps(target, { position: DEFAULT + 'remove' })
+    expect(target.position.toArray()).toStrictEqual([0, 0, 0])
+
+    // Recreate from instance args
+    instance.props.args = [1, 2, 3]
+    applyProps(target, { position: DEFAULT + 'remove' })
+    expect(target.position.toArray()).toStrictEqual([1, 2, 3])
+  })
```

Comment:
@drcmda, where are parent object args relevant when resetting for HMR? I have it prioritize like: `prop.args => object.args => []`.

Reply:
Dropping the first case, preferring to pick from a copy of the target object.